src/tools/genpolicy/src/registry.rs

```diff
@@ -252,130 +256,166 @@ async fn get_image_layers(
+                "Failed to create verity hash for {layer_digest}, error {e}"
+            );
+            error = true
+        };
+
+        if !error {
+            match get_verity_hash_value(&decompressed_path) {
+                Err(e) => {
+                    error_message = format!("Failed to get verity hash {e}");
+                }
+                Ok(v) => {
+                    verity_hash = v;
+                    if use_cached_files {
+                        add_verity_to_store(&cache_file, &layer_digest, &verity_hash)?;
+                    }
+                    info!("dm-verity root hash: {verity_hash}");
+                }
             }
         }
     }
 
-    if !use_cached_files {
-        let _ = std::fs::remove_dir_all(&base_dir);
-    } else if error {
-        delete_files(&decompressed_path, &compressed_path, &verity_path);
-    }
-
+    temp_dir.close()?;
     if error {
-        panic!("{error_message}");
+        if Path::new(cache_file).exists() {
+            std::fs::remove_file(&cache_file)?;
+        }
+        panic!("{}", &error_message);
     } else {
         Ok(verity_hash)
     }
 }
 
-async fn create_verity_hash_file(
-    use_cached_files: bool,
-    client: &mut Client,
-    reference: &Reference,
+#[derive(Serialize, Deserialize)]
+struct Layer {
+    layer_digest: String,
+    verity_hash: String,
+}
+
+// the store is a json file that matches layer hashes to verity hashes
+fn add_verity_to_store(
+    cache_file: &str,
     layer_digest: &str,
-    base_dir: &Path,
-    decompressed_path: &Path,
-    compressed_path: &Path,
-    verity_path: &Path,
+    verity_hash: &str,
 ) -> Result<()> {
-    if use_cached_files && decompressed_path.exists() {
-        info!("Using cached file {:?}", &decompressed_path);
+    // open the json file in read mode, create it if it doesn't exist
+    let read_file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(cache_file)?;
+
+    // If the file is empty, start with an empty JSON vector
+    let mut data: Vec<Layer> = if read_file.metadata()?.len() == 0 {
+        Vec::new()
     } else {
-        std::fs::create_dir_all(&base_dir)?;
+        let reader = BufReader::new(read_file);
+        serde_json::from_reader(reader)?
+    };
 
-        create_decompressed_layer_file(
-            use_cached_files,
-            client,
-            reference,
-            layer_digest,
-            &decompressed_path,
-            &compressed_path,
-        )
-        .await?;
+    // Add new data to the deserialized JSON
+    data.push(Layer{
+        layer_digest: layer_digest.to_string(),
+        verity_hash: verity_hash.to_string(),
+    });
+
+    // Serialize in pretty format
+    let serialized = serde_json::to_string_pretty(&data)?;
+
+    // Open the JSON file to write
+    let file = OpenOptions::new()
+        .write(true)
+        .open(cache_file)?;
+
+    let _ = file.lock_exclusive();
```

Comment:
Shouldn't this be:

if file.lock_exclusive().is_ok() {
    // Write the serialized JSON to the file
} else {
   // This verity hash will be cached when/if it gets calculated again in the future.
}

Reply:
the docs say `lock_exclusive()` is blocking so either the process can wait its turn, or we can skip over it and cache the next time around. Either is fine with me, let me know what you think.